crates/file_finder/src/file_finder.rs

```diff
@@ -256,47 +289,66 @@ impl Matches {
                 } else {
                     query_match.flatten()
                 };
-                Some((i + 1, Match::History(history_item.clone(), query_match)))
+                Some(Match::History(
+                    HistoryEntry {
+                        path: history_item.clone(),
+                        panel_match: query_match,
+                    },
+                    i + 1,
+                ))
             });
 
-        let mut unique_matches = BTreeSet::new();
-        self.matches = old_search_matches
-            .chain(history_matches)
-            .chain(new_search_matches)
-            .filter(|(_, m)| unique_matches.insert(m.clone()))
-            .sorted_by(|(history_score_a, a), (history_score_b, b)| {
-                match (a, b) {
-                    // bubble currently opened files to the top
-                    (Match::History(path, _), _) if Some(path) == currently_opened => {
-                        cmp::Ordering::Less
-                    }
-                    (_, Match::History(path, _)) if Some(path) == currently_opened => {
-                        cmp::Ordering::Greater
+        self.opened_path = currently_opened.cloned();
+
+        for new_match in history_matches.chain(new_search_matches) {
+            match self.matches.binary_search_by(|m| {
+                Self::cmp_matches(self.separate_history, currently_opened, &m, &new_match)
+            }) {
+                Ok(_duplicate) => continue,
+                Err(i) => {
+                    self.matches.insert(i, new_match);
+                    if self.matches.len() == 100 {
+                        break;
                     }
+                }
+            }
+        }
+    }
 
-                    (Match::History(_, _), Match::Search(_)) if self.separate_history => {
-                        cmp::Ordering::Less
-                    }
-                    (Match::Search(_), Match::History(_, _)) if self.separate_history => {
-                        cmp::Ordering::Greater
-                    }
+    fn cmp_matches(
+        separate_history: bool,
+        currently_opened: Option<&FoundPath>,
+        a: &Match,
+        b: &Match,
+    ) -> cmp::Ordering {
+        if a.eq(b) {
```

Comment:
Is it really needed? If we remove that, are those new `PartialEq` + `Eq` impls needed?
The `match` below seems to handle all that except `currently_opened` case.

Reply:
It seems like the confusion around all this comparison stuff for matches rised after we've merged history score into the match. And it's a bit confusing, since we want to compare for equality(i.e. when finding duplicate matches) without history score, but when we order the set of unique matches, we need the history score.

So this eq just compares matches without taking the history scores into account.

Reply:
Yes, looks a bit weird but at least more clear now.

It seems that CI fails and requires `itertools` to be removed from `file_finder/Cargo.toml` as we do not use it now, so we can add a brief commend and maybe just do `a == b` check instead?

Reply:
Now I think, maybe we do not need to store the history score with the match at all?


The reason why I've added it to the Match is to use the binary search to find the previously selected match in the new matches set. But since we want to find the particular match, it seems, that we should be able to do it passing 0 as a history score, having this check for equality in place.

If my understanding is correct, this would allow us to use the derived implementation for PartialEq.

Does it make sense?

Reply:
Less syntaxic noise seems always interesting to try.

I would say, we can even go and first binary_search inside of a new `Vec` where we insert `(Match, score)`, and then do `self.search_matches = that_new_vec.into_iter().map(..).collect()` later.

All those custom comparisons are quite confusing in that, already bloated, comparison logic.

Reply:
> I would say, we can even go and first binary_search inside of a new Vec where we insert (Match, score), and then do self.search_matches = that_new_vec.into_iter().map(..).collect() later.

This idea is not quite clear to me, could you please elaborate?

As of now we have retained old matches in `self.search_matches` and a couple of iterators over the new (unsorted AFAIU) matches collections. And then we add new matches to the `self.search_matches` keeping it sorted according to `cmp_matches`. Then, when we have sorted matches collection, we search for the previously selected match, since we want not only to check if it's still present, but also get it's new index to make it visible during the rendering.

Perhaps we could collect new matches in a vector, eliminating duplicates, then merge this new vector with the old matches. Though I'm not sure that creating this temporary vector is worth the effort, while we can do pretty much the same thing, but without extra copying.

Reply:
I think we are overthinking it, given how close are we to merge this 🙂 

Whatever way we can find to remove `eq` is nice, I think that's achieved by moving away the `usize`/history_score out of the struct (as you've noted well too).

I've tried to propose a way for doing that, as to first create an intermediate `Vec<(usize, Match)>` with whatever sorting/deduplication done that needs to know about `usize`, and then convert that into whatever `self.matches: Vec<Match>` type is by that time.
Then, presumably, we can remove the wrapper struct that was created for storing that `usize` only (as you were noting), and the diff will be even smaller.

But I do not insist on that way, if any other helps us to remove the odd `eq`.
I guess I can even live with that after some convincing, but feels quite easy to fix?